### ISSUE TITLE
Support large files

### ISSUE THREAD

**Issue by ebremont** [GrokImageCompression/grok#97]
The decoding process for large files (for example, about 33000 x 43000 pixels images) simply crashes ("Process stopped").
Here's the command line : 
./opj_decompress -i ../../big.JP2 -i /home/users/big.png -v
Here are the associated logs :
[INFO] Start to read j2k main header (3223)
[INFO] Main header has been correctly decoded.
[INFO] No decoded area parameters, set the decoded area to the whole image
[INFO] Header of tile 1 / 1 has been read.
Process stopped

Same results when trying to decode a little part of the image (<0,0,100,100>)
The decoding process with Kakadu seems to be ok.

**Comment by boxerab** [comment]
Thanks. Can you share a sample file? I realize it will be big!

**Comment by boxerab** [comment]
Would this image work?

http://www.skyandtelescope.com/astronomy-news/a-half-gigabyte-view-ofthemoon/

**Comment by boxerab** [comment]
Also, please tell us more about your system and OS.

**Comment by ebremont** [comment]
Unfortunately, I can't give you the sample files because there are under licence. But I can give you the main main header structure or make all the tests you want :-)
However, I am going to test the given file this afternoon.
About the os: the exec is running on a CentOs7 with 8Go ram. 
Is it possible to add more logs ?

**Comment by retokromer** [comment]
Side note: as much RAM as possible is always a good thing when working with AV files!

**Comment by boxerab** [comment]
Yes, you will need more RAM to handle such a large image.

**Comment by boxerab** [comment]
Please either share an image that is crashing, or try on a system with more RAM and let us know if it crashes.  We use 32 bit buffers to decompress to, so that is already about 5 GB, and we need to allocate for the compressed JP2 image as well, so it looks like you don't have enough memory.  

Agree that logging should be better.

**Comment by ebremont** [comment]
As Kakadu can decode this with a 2Go ram computer, I hoped 8Go ram is OK. 
I can try with 16go.
Does grok has memory allocation optimizations ?

**Comment by boxerab** [comment]
Yes, it has optimizations, but of course there is always more that can be done.  Perhaps I can log how much memory is being allocated (in verbose mode) so that it will be easier to trouble shoot this.

**Comment by ebremont** [comment]
Just for precisions : I work only on 64bits system.
Can I supposed that 10Go RAM will be sufficient ? 
I have some difficulties to find a 16Go computer. I hope that I will be able to make tests on a very good machine in about 2 hours. 
Howerver, the target application os will be a Cento6 64 bits with 16Go (for application + image decoder). I really hope that Grok will be able to read our big jpeg2000 files.
I have no really constraints on cpu or time, but I have RAM limitations. :(

**Comment by boxerab** [comment]
So, with latest checkin, you should see console output if you are indeed running out of memory.

**Comment by ebremont** [comment]
Ok. Thanks !! I get your version, I compile as soon as possible. :-)
Latest news : I succeed decoding the image on a very powerful pc with the same OS. The decode process ends successfully but seems to need about 18Go RAM. :(  Too much for me (15Go max)...
Decoding a little part of the image seems to require more memory.
Is 3Go RAM optimization feasible? 

NB : I don't know if it can help, but I attached the main header description. 
[main_header.zip](https://github.com/GrokImageCompression/grok/files/1165845/main_header.zip)

**Comment by boxerab** [comment]
Well, this makes sense:  2 components @ 33000x43000 resolution using 32 bit decode
equals over 10 GB of memory. 

It is possible to break the image into stripes and process stripe by stripe, but this is a lot of work on the code - don't know if/when that would be possible.

**Comment by ebremont** [comment]
Uhm, ok. Thanks for your answer :-)
My questions seem probably stupid, but does the decoded output file format have an impact on memory performances ? Is this possible to decode the image and write the decoded image in many little files (except by calling the decoding process many times with -d option) ?

**Comment by boxerab** [comment]
One thing we could do would be to decode component by component. This would halve the memory usage for a two component image.   Also, it might be possible to use 16 bit fixed-point instead of 32 bit fixed point.  I think this is what Kakadu does.

**Comment by boxerab** [comment]
I see the image has 15 layers, Do you need to decode all of the layers?

**Comment by boxerab** [comment]
Or you could try a lower resolution.

**Comment by ebremont** [comment]
Kakadu (kdu_buffered_expand) seems to use about 120Mo RAM for the same image.
If it is possible to have an option to decode component by component, and also to reduce memory need (even if decoding process is longer) , I will be great. :-)
For the resolution, in theory, I am supposed to decode the image without deterioration...  I just can "split" in smaller pieces :-( But, you're right, I will ask what's the target resolution to see if downgrading the resolution is possible.

**Comment by boxerab** [comment]
I just realized that I can't decode component by component, because your image uses the multi-component transform, so all three decoded YUV components must be present in memory to convert to RGB.

You can pick up 32 GB of DDR4 memory for around USD $200, so I would think this is the way to go
for dealing with these images. It will take me a long time to get memory usage reduced close to Kakadu, 

Can you tell me more about what kind of images these are, and how you are using them ?

**Comment by boxerab** [comment]
I'm going to close this issue.  The library can read large images, it just needs enough system RAM to do this.

**Comment by boxerab** [comment]
This is a duplicate of issue #41 . Feel free to add comments to that issue if you like.

**Comment by boxerab** [comment]
Thanks for reporting this, by the way :)

**Comment by ebremont** [comment]
I have a 16Go RAM limitation, not on account of budget constraints, but on account of project constraints. I succeeded testing on a powerful platform, but it is only for test and development.
Also, I can't use on my project 20Go to decode a 130Mo image... (These images are geo-referenced satellite pictures).

I understand that you closed this issue. But I think you should note this point as a future enhancement of your software, so as to allow persons, as me, who think that this software would be a good alternative to Kakadu.

**Comment by boxerab** [comment]
Hi Estelle,
I'm sorry the library won't work for your project.  Please keep in mind that Kakadu is funded by commercial licenses costing thousands of dollars, while Grok is freely available and developed in my spare time. 

Also, as I commented above, issue #41, which remains open, addresses the issue of memory usage.
So, in the future, it may be addressed.

Kind Regards,
Aaron

**Comment by ebremont** [comment]
Hi Aaron, 
I respect your work, and the work of this project contributors.
I think this project is a good initiative, and proposes a concrete alternative to others decoders. Kakadu is very expensive because it relies on about 15 years experience. I tried so many open sources and different softwares.
I've just tried to contribute here by sharing some constraints by reading some files encoded with Kakadu, because I truly believe that this project has a really good potential and is a good work.
Good continuation !
Kind regards, 
Estelle
